mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala

```diff
@@ -55,31 +48,59 @@ class MulticlassMetrics @Since("1.1.0") (predictionAndLabels: RDD[_ <: Product])
         throw new IllegalArgumentException(s"Expected Row of tuples, got $other")
     })
 
-  private lazy val labelCountByClass: Map[Double, Double] =
-    predLabelsWeight.map {
-      case (_: Double, label: Double, weight: Double) =>
-        (label, weight)
-    }.reduceByKey(_ + _)
+
+  private val confusions = predictionAndLabels.map {
+    case (prediction: Double, label: Double, weight: Double) =>
+      (prediction, label, weight)
+    case (prediction: Double, label: Double) =>
+      (prediction, label, 1.0)
+    case other =>
+      throw new IllegalArgumentException(s"Expected tuples, got $other")
+  }.map { case (prediction: Double, label: Double, weight: Double) =>
+    ((label, prediction), weight)
+  }.reduceByKey(_ + _)
     .collectAsMap()
+
+  private lazy val labelCountByClass: Map[Double, Double] = {
```

Comment:
OK yeah I was also thinking you could just compute these all non-lazily, and do so in one pass over confusions, but, maybe not worth it.